runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/splittabledofn/SDFFeederViaStateAndTimers.java

```diff
@@ -0,0 +1,179 @@
+  private ValueState<WindowedValue<KV<InputT, RestrictionT>>> seedState;
+
+  private final StateTag<ValueState<RestrictionT>> restrictionTag;
+  private ValueState<RestrictionT> restrictionState;
+
+  private StateTag<WatermarkHoldState> watermarkHoldTag =
+      StateTags.makeSystemTagInternal(
+          StateTags.<GlobalWindow>watermarkStateInternal("hold", TimestampCombiner.LATEST));
+  private WatermarkHoldState holdState;
+
+  private Instant inputTimestamp;
+  private BundleSplit split;
+
+  /** Initializes the feeder. */
+  public SDFFeederViaStateAndTimers(
+      StateInternals stateInternals,
+      TimerInternals timerInternals,
+      Coder<InputT> elementWireCoder,
+      Coder<RestrictionT> restrictionWireCoder,
+      Coder<BoundedWindow> windowCoder) {
+    this.stateInternals = stateInternals;
+    this.timerInternals = timerInternals;
+    this.windowCoder = windowCoder;
+    this.elementRestrictionWireCoder =
+        FullWindowedValueCoder.of(KvCoder.of(elementWireCoder, restrictionWireCoder), windowCoder);
+    this.seedTag = StateTags.value("seed", elementRestrictionWireCoder);
+    this.restrictionTag = StateTags.value("restriction", restrictionWireCoder);
+  }
+
+  /** Passes the initial element/restriction pair. */
+  public void seed(WindowedValue<KV<InputT, RestrictionT>> elementRestriction) {
+    initState(
+        StateNamespaces.window(
+            windowCoder, Iterables.getOnlyElement(elementRestriction.getWindows())));
+    seedState.write(elementRestriction);
+    inputTimestamp = elementRestriction.getTimestamp();
+  }
+
+  /**
+   * Resumes from a timer and returns the current element/restriction pair (with an up-to-date value
+   * of the restriction).
+   */
+  public WindowedValue<KV<InputT, RestrictionT>> resume(TimerData timer) {
+    initState(timer.getNamespace());
+    WindowedValue<KV<InputT, RestrictionT>> seed = seedState.read();
+    inputTimestamp = seed.getTimestamp();
+    return seed.withValue(KV.of(seed.getValue().getKey(), restrictionState.read()));
+  }
+
+  /**
+   * Commits the state and timers: clears both if no checkpoint happened, or adjusts the restriction
+   * and sets a wake-up timer if a checkpoint happened.
+   */
+  public void commit() throws IOException {
+    if (split == null) {
+      // No split - the call terminated.
+      seedState.clear();
+      restrictionState.clear();
+      holdState.clear();
+      return;
+    }
+
+    // For now can only happen on the first instruction which is SPLITTABLE_PROCESS_ELEMENTS.
+    List<DelayedApplication> residuals = split.getResidualRootsList();
+    checkArgument(residuals.size() == 1, "More than 1 residual is unsupported for now");
+    DelayedApplication residual = residuals.get(0);
+
+    ByteString encodedResidual = residual.getApplication().getElement();
+    WindowedValue<KV<InputT, RestrictionT>> decodedResidual =
+        CoderUtils.decodeFromByteArray(elementRestrictionWireCoder, encodedResidual.toByteArray());
```

Comment:
nit: use the InputStream available on ByteStrings
`elementRestrictionWireCoder.decode(encodedResidual.newInput())`